design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,214 @@
+and generate the documentation files using the operator observability docs
+utilities.
+
+### Declarative configuration
+
+The declarative configuration will be done using a YAML files. The YAML files
+will be parsed and used to create the metrics, alerts, and recording rules.
+Components should be able to define their resources in one or more YAML files.
+
+As a general rule, the files will be structured in the following way:
+
+```yaml
+observability:
+  common_labels:
+    - name: controller
+      value: example-operator
+
+  groups:
+    - name: operator_resources
+      common_labels:
+        - name: grouped
+          value: this_is_grouped
+      metrics:
+        - name: example_operator_reconcile_count
+          help: Number of times the operator has executed the reconcile loop
+          type: counter
+          stability: stable
+          labels:
+            - name: another_label
+              value: only_for_this_metric
+
+      recording_rules:
+        - name: example_operator_number_of_pods
+          help: Number of example operator pods in the cluster
+          type: gauge
+          expr: sum(up{namespace='$NAMESPACE', pod=~'example-operator-.*'}) or vector(0)
+      
+      alerts:
+        - name: OperatorDown
+          expr: example_operator_number_of_pods == 0
+          for: 5m
+          annotations:
+            summary: Example operator is down
+            description: Example operator is down for more than 5 minutes.
+            runbook_url: $RUNBOOK_BASE_URL/OperatorDown
+          labels:
+            severity: critical
+```
+
+But components can have a single file only for metrics or for a specific group
+of metrics, for example:
+
+```yaml
+observability:
+   groups:
+      - name: operator_resources
+        common_labels:
+           - name: grouped
+             value: this_is_grouped
+        metrics:
+           - name: example_operator_reconcile_count
+             help: Number of times the operator has executed the reconcile loop
+             type: counter
+             stability: stable
+             labels:
+                - name: another_label
+                  value: only_for_this_metric
+
+      - name: other_group
+        metrics:
+          ...
+```
+
+In the end, all the declarative configuration will be parsed and joined into a
+single `Observability` struct, which will be used to create the metrics, alerts,
+and recording rules. The same idea will be applied to other resources.
+
+It should be possible to:
+
+- create a single file with all groups of metrics, alerts, and recording rules
+
+- create multiple files, each with a specific set of groups. With each group
+containing its metrics, alerts, and recording rules
+
+- create multiple files, each with multiple groups with only one type of
+resource (metrics, alerts, or recording rules). For example, one file with all
+metrics, another with all alerts, and another with all recording rules.
```

Comment:
I agree, This would be a great improvement.
We need to enable an allow/deny list .

But, I dont think we should have metrics, alerts, and recording rules in a single file.
And we should have the structure for having an allow/deny list fro what we expose.